File: go/request/rtb_twofive/native.go
package twofive

import "github.com/francoispqt/gojay"

// Native object represents a native type impression. Native ad units are intended to blend seamlessly into
// the surrounding content (e.g., a sponsored Twitter or Facebook post). As such, the response must be
// well-structured to afford the publisher fine-grained control over rendering.
type Native struct {
	ID      string `json:"id,omitempty"      valid:"optional"`
	Request string `json:"request,omitempty" valid:"required"`
	Ver     string `json:"ver,omitempty"     valid:"optional"`
	API     []int  `json:"api,omitempty"     valid:"optional"`
	Battr   []int  `json:"battr,omitempty"   valid:"optional"`
}

// MarshalJSONObject implements MarshalerJSONObject
func (n *Native) MarshalJSONObject(enc *gojay.Encoder) {
	enc.StringKeyOmitEmpty("id", n.ID)
	enc.StringKeyOmitEmpty("request", n.Request)
	enc.StringKeyOmitEmpty("ver", n.Ver)
	var aPISlice = Ints(n.API)
	enc.ArrayKeyOmitEmpty("api", aPISlice)
	var battrSlice = Ints(n.Battr)
	enc.ArrayKeyOmitEmpty("battr", battrSlice)
}

// IsNil checks if instance is nil
func (n *Native) IsNil() bool {
	return n == nil
}

// UnmarshalJSONObject implements gojay's UnmarshalerJSONObject
func (n *Native) UnmarshalJSONObject(dec *gojay.Decoder, k string) error {

	switch k {
	case "id":
		return dec.String(&n.ID)

	case "request":
		return dec.String(&n.Request)

	case "ver":
		return dec.String(&n.Ver)

	case "api":
		var aSlice = Ints{}
		err := dec.Array(&aSlice)
		if err == nil && len(aSlice) > 0 {
			n.API = []int(aSlice)
		}
		return err

	case "battr":
		var aSlice = Ints{}
		err := dec.Array(&aSlice)
		if err == nil && len(aSlice) > 0 {
			n.Battr = []int(aSlice)
		}
		return err

	}
	return nil
}

// NKeys returns the number of keys to unmarshal
func (n *Native) NKeys() int { return 0 }

File: kotlin/src/commonMain/kotlin/com/adsbynimbus/openrtb/request/Native.kt
package com.adsbynimbus.openrtb.request

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.jvm.JvmField

/**
 * This public class represents a native type impression.
 *
 * Native ad units are intended to blend seamlessly into the surrounding content (e.g., a sponsored
 * Twitter or Facebook post). As such, the response must be well-structured to afford the publisher
 * fine-grained control over rendering. The Native Subcommittee has developed a companion
 * specification to OpenRTB called the Dynamic Native Ads API. It defines the request parameters and
 * response markup structure of native ad units. This object provides the means of transporting
 * request parameters as an opaque string so that the specific parameters can evolve separately
 * under the auspices of the Dynamic Native Ads API. Similarly, the ad markup served will be
 * structured according to that specification. The presence of a Native as a subordinate of the Imp
 * object indicates that this impression is offered as a native type impression. At the publisher’s
 * discretion, that same impression may also be offered as banner, video, and/or audio by also
 * including as Imp subordinates objects of those types. However, any given bid for the impression
 * must conform to one of the offered types.
 *
 * [OpenRTB Section 3.2.9](https://www.iab.com/wp-content/uploads/2016/03/OpenRTB-API-Specification-Version-2-5-FINAL.pdf#page=23)
 *
 * @property bidfloor Minimum bid for this native impression expressed in CPM.
 * @property request Request payload complying with the Native Ad Specification
 * @property ver Version of the Dynamic Native Ads API to which request complies. Highly recommended
 *               for efficient parsing.
 * @property api List of supported API frameworks for this impression. If an API is not explicitly
 *               listed, it is assumed not to be supported.
 * @property battr Set of creative attributes to block.
 */
@Serializable
public class Native(
    @JvmField @SerialName("bidfloor") public val bidfloor: Float? = null,
    @JvmField @SerialName("request") public val request: String,
    @JvmField @SerialName("ver") public val ver: String = "",
    @JvmField @SerialName("api") public val api: ByteArray = ByteArray(0),
    @JvmField @SerialName("battr") public val battr: ByteArray = ByteArray(0),
)
